docs/src/explanations/properties.md

```diff
@@ -148,3 +148,249 @@ on integral `Property` typed values.
+  val descriptions = IO(Output(Property[Seq[csrDescType.Type]]()))
+
+  // Instantiate a couple CSR modules.
+  val mcycle = Module(new CSRModule(csrDescDef, 64, "mcycle", "Machine cycle counter."))
+  val minstret = Module(new CSRModule(csrDescDef, 64, "minstret", "Machine instructions-retired counter."))
+
+  // Assign references to the CSR description objects to the property port.
+  descriptions := Property(Seq(mcycle.description.as(csrDescType), minstret.description.as(csrDescType)))
+}
+```
+
+The `Top` module represents the entrypoint. It creates the `Definition` of the
+`CSRDescription`, and creates some `CSRModule`s. It then takes the description
+references, collects them into a list, and outputs the list so it will be
+exposed to the outside.
+
+To illustrate what this example generates, here is a listing of the FIRRTL:
+
+```
+FIRRTL version 4.0.0
+circuit Top :
+  class CSRDescription :
+    output identifier : String
+    output description : String
+    output width : Integer
+    input identifierIn : String
+    input descriptionIn : String
+    input widthIn : Integer
+
+    propassign identifier, identifierIn
+    propassign description, descriptionIn
+    propassign width, widthIn
+
+  module mcycle :
+    input clock : Clock
+    input reset : Reset
+    output value : UInt<64>
+    output description : Inst<CSRDescription>
+
+    object csrDescription of CSRDescription
+    propassign csrDescription.identifierIn, String("mcycle")
+    propassign csrDescription.descriptionIn, String("Machine cycle counter.")
+    propassign csrDescription.widthIn, Integer(64)
+    regreset csr : UInt<64>, clock, reset, UInt<64>(0h0)
+    connect value, csr
+    propassign description, csrDescription
+
+  module minstret :
+    input clock : Clock
+    input reset : Reset
+    output value : UInt<64>
+    output description : Inst<CSRDescription>
+
+    object csrDescription of CSRDescription
+    propassign csrDescription.identifierIn, String("minstret")
+    propassign csrDescription.descriptionIn, String("Machine instructions-retired counter.")
+    propassign csrDescription.widthIn, Integer(64)
+    regreset csr : UInt<64>, clock, reset, UInt<64>(0h0)
+    connect value, csr
+    propassign description, csrDescription
+
+  public module Top :
+    input clock : Clock
+    input reset : UInt<1>
+    output descriptions : List<Inst<CSRDescription>>
+
+    inst mcycle of mcycle
+    connect mcycle.clock, clock
+    connect mcycle.reset, reset
+    inst minstret of minstret
+    connect minstret.clock, clock
+    connect minstret.reset, reset
+    propassign descriptions, List<Inst<CSRDescription>>(mcycle.description, minstret.description)
+```
+
+To understand the `Object` graph that is constructed, we will consider an
+entrypoint to elaboration, and then show a hypothetical JSON representation of
+the `Object` graph. The details of how we go from IR to an `Object` graph are
```

Comment:
Currently, we don't have the API to get the hypothetical JSON representation. we can demo it in the panamaom framework.

Reply:
Yeah that would be great. I will update the wording here slightly, since panamaom isn't an "external tool". Maybe we can add an example in this section of the doc showing a concrete use of panamaom with this IR. It could traverse the descriptions and print something out, so we could show the code snippet and the expected output, or something like that.